### PR TITLE
specific error for (lag|lead)(default=) when incompatible type.

### LINKED ISSUE #5073: Error when default value in lag() and lead() is of different class
Using a default value in `lag()` and `lead()` from the devel version of dplyr throws an error if the value has a different class from the values in `x` (it did not use to throw this error). I could not see this change in NEWS.md (but I might have missed it).

```r
lag(c("1", "2", "3"), default = FALSE)
# Error: No common type for `..1` <logical> and `..2` <character>.
# This used to work.
```

```
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.6.2 (2019-12-12)
 os       macOS Catalina 10.15.3      
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language (EN)                        
 collate  en_GB.UTF-8                 
 ctype    en_GB.UTF-8                               

─ Packages ────────────────────────────────────────────────────────────────────────
 package     * version     date       lib source                          
 assertthat    0.2.1       2019-03-21 [1] CRAN (R 3.6.0)                  
 backports     1.1.5       2019-10-02 [1] CRAN (R 3.6.0)                  
 broom         0.5.5       2020-02-29 [1] CRAN (R 3.6.0)                  
 callr         3.4.2       2020-02-12 [1] CRAN (R 3.6.0)                  
 cellranger    1.1.0       2016-07-27 [1] CRAN (R 3.6.0)                  
 cli           2.0.2       2020-02-28 [1] CRAN (R 3.6.0)                  
 colorspace    1.4-1       2019-03-18 [1] CRAN (R 3.6.0)                  
 crayon        1.3.4       2017-09-16 [1] CRAN (R 3.6.0)                  
 DBI           1.1.0       2019-12-15 [1] CRAN (R 3.6.0)                  
 dbplyr        1.4.2       2019-06-17 [1] CRAN (R 3.6.0)                  
 desc          1.2.0       2018-05-01 [1] CRAN (R 3.6.0)                  
 devtools      2.2.2       2020-02-17 [1] CRAN (R 3.6.0)                  
 digest        0.6.25      2020-02-23 [1] CRAN (R 3.6.2)                  
 dplyr       * 0.8.99.9002 2020-03-30 [1] Github (tidyverse/dplyr@eb4d4f3)
 ellipsis      0.3.0       2019-09-20 [1] CRAN (R 3.6.0)                  
 fansi         0.4.1       2020-01-08 [1] CRAN (R 3.6.0)                  
 forcats     * 0.5.0       2020-03-01 [1] CRAN (R 3.6.0)                  
 fortunes      1.5-4       2016-12-29 [1] CRAN (R 3.6.0)                  
 fs            1.3.2       2020-03-05 [1] CRAN (R 3.6.2)                  
 generics      0.0.2       2018-11-29 [1] CRAN (R 3.6.0)                  
 ggplot2     * 3.3.0       2020-03-05 [1] CRAN (R 3.6.2)                  
 glue          1.3.2       2020-03-12 [1] CRAN (R 3.6.0)                  
 gtable        0.3.0       2019-03-25 [1] CRAN (R 3.6.0)                  
 haven         2.2.0       2019-11-08 [1] CRAN (R 3.6.0)                  
 hms           0.5.3       2020-01-08 [1] CRAN (R 3.6.0)                  
 httr          1.4.1       2019-08-05 [1] CRAN (R 3.6.0)                  
 jsonlite      1.6.1       2020-02-02 [1] CRAN (R 3.6.0)                  
 lattice       0.20-40     2020-02-19 [1] CRAN (R 3.6.0)                  
 lifecycle     0.2.0       2020-03-06 [1] CRAN (R 3.6.2)                  
 lubridate     1.7.4       2018-04-11 [1] CRAN (R 3.6.0)                  
 magrittr      1.5         2014-11-22 [1] CRAN (R 3.6.0)                  
 memoise       1.1.0       2017-04-21 [1] CRAN (R 3.6.0)                  
 modelr        0.1.6       2020-02-22 [1] CRAN (R 3.6.2)                  
 munsell       0.5.0       2018-06-12 [1] CRAN (R 3.6.0)                  
 nlme          3.1-145     2020-03-04 [1] CRAN (R 3.6.2)                  
 packrat       0.5.0       2018-11-14 [1] CRAN (R 3.6.0)                  
 pillar        1.4.3       2019-12-20 [1] CRAN (R 3.6.0)                  
 pkgbuild      1.0.6       2019-10-09 [1] CRAN (R 3.6.0)                  
 pkgconfig     2.0.3       2019-09-22 [1] CRAN (R 3.6.0)                  
 pkgload       1.0.2       2018-10-29 [1] CRAN (R 3.6.0)                  
 prettyunits   1.1.1       2020-01-24 [1] CRAN (R 3.6.0)                  
 processx      3.4.2       2020-02-09 [1] CRAN (R 3.6.2)                  
 ps            1.3.2       2020-02-13 [1] CRAN (R 3.6.0)                  
 purrr       * 0.3.3       2019-10-18 [1] CRAN (R 3.6.0)                  
 R6            2.4.1       2019-11-12 [1] CRAN (R 3.6.0)                  
 Rcpp          1.0.4       2020-03-17 [1] CRAN (R 3.6.0)                  
 readr       * 1.3.1       2018-12-21 [1] CRAN (R 3.6.0)                  
 readxl        1.3.1       2019-03-13 [1] CRAN (R 3.6.0)                  
 remotes       2.1.1       2020-02-15 [1] CRAN (R 3.6.0)                  
 reprex        0.3.0       2019-05-16 [1] CRAN (R 3.6.0)                  
 rlang         0.4.5.9000  2020-03-30 [1] Github (r-lib/rlang@a90b04b)    
 rprojroot     1.3-2       2018-01-03 [1] CRAN (R 3.6.0)                  
 rstudioapi    0.11        2020-02-07 [1] CRAN (R 3.6.0)                  
 rvest         0.3.5       2019-11-08 [1] CRAN (R 3.6.0)                  
 scales        1.1.0       2019-11-18 [1] CRAN (R 3.6.0)                  
 sessioninfo   1.1.1       2018-11-05 [1] CRAN (R 3.6.0)                  
 stringi       1.4.6       2020-02-17 [1] CRAN (R 3.6.0)                  
 stringr     * 1.4.0       2019-02-10 [1] CRAN (R 3.6.0)                  
 testthat      2.3.2       2020-03-02 [1] CRAN (R 3.6.2)                  
 tibble      * 3.0.0       2020-03-30 [1] CRAN (R 3.6.2)                  
 tidyr       * 1.0.2       2020-01-24 [1] CRAN (R 3.6.0)                  
 tidyselect    1.0.0       2020-01-27 [1] CRAN (R 3.6.0)                  
 tidyverse   * 1.3.0       2019-11-21 [1] CRAN (R 3.6.0)                  
 usethis       1.5.1       2019-07-04 [1] CRAN (R 3.6.0)                  
 vctrs         0.2.99.9010 2020-03-30 [1] Github (r-lib/vctrs@65849f9)    
 withr         2.1.2       2018-03-15 [1] CRAN (R 3.6.0)                  
 xml2          1.2.5       2020-03-11 [1] CRAN (R 3.6.0)
```

### PR DESCRIPTION
**Pull request by romainfrancois**
close #5073

``` r
library(dplyr, warn.conflicts = FALSE)

lag(c("1", "2", "3"), default = FALSE)
#> Error: Incompatible type for `default`.
#> ℹ <logical> cannot be cast to type <character>.
```

<sup>Created on 2020-03-31 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

File: R/lead-lag.R [hunk 1]
```diff
@@ -63,6 +78,7 @@ lag <- function(x, n = 1L, default = NA, order_by = NULL, ...) {
   xlen <- vec_size(x)
   n <- pmin(n, xlen)
 
+  default <- leadlag_default(x, default)
```
Comment (hadley) [inline R/lead-lag.R]: Could we just do `default <- vec_cast(default, x, x_arg = "default", to_arg = "x")`? I would prefer not to invest a bunch of time into the error messages of these functions at the moment.

**Comment by hadley** [comment]
We also need a bullet in the news stating that `lead()`/`lag()` are stricter about their inputs.

File: R/lead-lag.R (hunk 1 above)
Comment (DavisVaughan) [reply R/lead-lag.R]: I agree with this

Comment (romainfrancois) [reply R/lead-lag.R]: Of course. Thanks both.

**Pull request merged**